### PR TITLE
player class implementation

### LINKED ISSUE #7: Implement Player class Task#7

### LINKED ISSUE #8: Implement method checking for available space for Player Task#8

### PR DESCRIPTION
**Pull request by ArturGubala**
closes #7 closes #8

Here is my implementation for class Player, and function which is responsible for counting free space on ship for cargo.

### CODE REVIEW

File: shm/Player.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+public:
+    Player(std::unique_ptr<Ship> ship, size_t money) {
+        ship_ = std::move(ship);
+        money_ = money;
+        avaliableSpace_ = countAvailableSpace();
+    };
+
+    size_t getMoney() const { return money_; };
+    size_t getAvaliableSpace() const { return avaliableSpace_; };
+    size_t getSpeed() const { return ship_->getSpeed(); };
```
Comment (J-Cieplinski) [inline shm/Player.hpp:18]: I think it should return const value

File: shm/Player.hpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <algorithm>
+#include <memory>
+#include <numeric>
```
Comment (J-Cieplinski) [inline shm/Player.hpp]: `<numeric>` and `<algorithm>` should be imported in cpp file. It's unnecessary here

**Review by J-Cieplinski — CHANGES_REQUESTED** [review]

File: shm/Player.hpp (hunk 1 above)
Comment (ArturGubala) [reply shm/Player.hpp:18]: You mean "return const ship_->getSpeed();"

File: shm/Player.hpp (hunk 2 above)
Comment (ArturGubala) [reply shm/Player.hpp]: Yeah, I tested my cod in one hpp file and forgot detete them. I will fix it. Thx!

File: shm/Player.hpp (hunk 1 above)
Comment (J-Cieplinski) [reply shm/Player.hpp:18]: I mean `const size_t getMoney()...` etc :)

Comment (ArturGubala) [reply shm/Player.hpp:18]: Hmm, I thought that "const" after name of getter protect us.

Comment (J-Cieplinski) [reply shm/Player.hpp:18]: const after the function name "promises" that the getter itself won't modify variables.

Comment (ArturGubala) [reply shm/Player.hpp:18]: In this case we return copy of values, so I think it isn't necessary to return const, but maybe I'm wrong.

File: shm/Player.cpp [hunk 3]
```diff
@@ -1 +1,11 @@
 #include "Player.hpp"
+
+size_t countAvailableSpace() {
+    auto total = std::accumulate(begin(ship_->getAllCargos()), end(ship_->getAllCargos()), 0, [](const auto& cargo) {
+        cargo->getAmount();
```
Comment (ArturGubala) [inline shm/Player.cpp]: Arrow operator is correct here? When I thinking about it now I should use dot operator.

File: shm/Player.cpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    });
+    if (ship_->getCapacity() <= total) {
+        return 0;
+    }
+    return ship_->getCapacity() - total;
+}
```
Comment (iprice73) [inline shm/Player.cpp]: new line here

File: shm/Player.hpp (hunk 1 above)
Comment (iprice73) [reply shm/Player.hpp:18]: On livestream they did not mention about const before get(), so I would stick with this

**Review by iprice73 — COMMENTED** [review]
Some minor changes, but looks fine :)

File: shm/Player.cpp (hunk 3 above)
Comment (J-Cieplinski) [reply shm/Player.cpp]: Hard to tell because there is no Ship implementation on our git 😄

File: shm/Player.hpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
```
Comment (lisie-sprawy) [inline shm/Player.hpp:2]: After #pragma once we should give empty spaces or it doesn't matter? After checking several student's PR I don't know.

File: shm/Player.hpp (hunk 1 above)
Comment (niemarcin) [reply shm/Player.hpp:18]: I think that for now it is enough if we have `type functionName() const {}` to ensure that our getters won't modify the variables. 
+1 to @LogarithmDerivative

**Review by lisie-sprawy — APPROVED** [review]
Looks good for me :)

File: shm/Player.cpp (hunk 3 above)
Comment (niemarcin) [reply shm/Player.cpp]: Actually it is, in module1/solutions/exercise3.cpp but not that much of it 😄  No getAllCargos() function nor cargo object are present there.

But basically it depends how Cargo will be stored in Ship. If by pointer then -> operator is fine.

File: shm/Player.hpp (hunk 5 above)
Comment (J-Cieplinski) [reply shm/Player.hpp:2]: I would go with one new line

File: shm/Player.cpp (hunk 3 above)
Comment (niemarcin) [reply shm/Player.cpp]: What do you think of changing name "total" to something different like "totalOccupied" or "occupiedSpace"? At first I thought that "total" is whole ship capacity.

**Review by niemarcin — COMMENTED** [review]
@ArturGubala please take a look at comments

**Review by J-Cieplinski — APPROVED** [review]

File: shm/Player.cpp (hunk 4 above)
Comment (ArturGubala) [reply shm/Player.cpp]: Fixed.

File: shm/Player.cpp (hunk 3 above)
Comment (ArturGubala) [reply shm/Player.cpp]: You're right that total is not the best name for this variable. From your two suggestions I  prefer "occupiedSpace", so I will change on it if everyone else agree.

File: shm/Player.hpp (hunk 5 above)
Comment (ArturGubala) [reply shm/Player.hpp:2]: Me too, but if I good remember Mateusz said that is just personal preferences.

**Review by niemarcin — COMMENTED** [review]
We would probably need to include Ship and Cargo header files as well.

File: shm/Player.cpp (hunk 3 above)
Comment (ArturGubala) [reply shm/Player.cpp]: Changed.

**Review by niemarcin — APPROVED** [review]

**Pull request merged**